mesh/pumi.cpp

```diff
@@ -0,0 +1,1909 @@
+   apf::Numbering* v_num_loc = apf::createNumbering(apf_mesh, "VertexNumbering",
+                                                    crd_shape, 1);
+   // Check if it is a curved mesh
+   curved = (crd_shape->getOrder() > 1) ? 1 : 0;
+
+   // Read mesh
+   ReadSCORECMesh(apf_mesh, v_num_loc, curved);
+   mfem::out << "After ReadSCORECMesh" << endl;
+   // at this point the following should be defined:
+   //  1) Dim
+   //  2) NumOfElements, elements
+   //  3) NumOfBdrElements, boundary
+   //  4) NumOfVertices, with allocated space in vertices
+   //  5) curved
+   //  5a) if curved == 0, vertices must be defined
+   //  5b) if curved != 0 and read_gf != 0,
+   //         'input' must point to a GridFunction
+   //  5c) if curved != 0 and read_gf == 0,
+   //         vertices and Nodes must be defined
+
+   // FinalizeTopology() will:
+   // - assume that generate_edges is true
+   // - assume that refine is false
+   // - does not check the orientation of regular and boundary elements
+   FinalizeTopology();
+
+   if (curved && read_gf)
+   {
+      // Check it to be only Quadratic if higher order
+      // mfem::out << "Is Curved?: "<< curved << "\n" <<read_gf <<endl;
+      Nodes = new GridFunctionPumi(this, apf_mesh, v_num_loc,
+                                   crd_shape->getOrder());
+      edge_vertex = NULL;
+      own_nodes = 1;
+      spaceDim = Nodes->VectorDim();
+      // if (ncmesh) { ncmesh->spaceDim = spaceDim; }
+      // Set the 'vertices' from the 'Nodes'
+      for (int i = 0; i < spaceDim; i++)
+      {
+         Vector vert_val;
+         Nodes->GetNodalValues(vert_val, i+1);
+         for (int j = 0; j < NumOfVertices; j++)
+         {
+            vertices[j](i) = vert_val(j);
+         }
+      }
+   }
+
+   // Delete numbering
+   apf::destroyNumbering(v_num_loc);
+
+   Finalize(refine, fix_orientation);
+}
+
+void PumiMesh::ReadSCORECMesh(apf::Mesh2* apf_mesh, apf::Numbering* v_num_loc,
+                              const int curved)
+{
+   // Here fill the element table from SCOREC MESH
+   // The vector of element pointers are generated with attr and connectivity
+
+   apf::MeshIterator* itr = apf_mesh->begin(0);
+   apf::MeshEntity* ent;
+   NumOfVertices = 0;
+   while ((ent = apf_mesh->iterate(itr)))
+   {
+      // ids start from 0
+      apf::number(v_num_loc, ent, 0, 0, NumOfVertices);
+      NumOfVertices++;
+   }
+   apf_mesh->end(itr);
+
+   Dim = apf_mesh->getDimension();
+   NumOfElements = countOwned(apf_mesh,Dim);
+   elements.SetSize(NumOfElements);
+
+   // Get the attribute tag
+   // apf::MeshTag* attTag = apf_mesh->findTag("attribute");
+
+   // read elements from SCOREC Mesh
+   itr = apf_mesh->begin(Dim);
+   unsigned int j=0;
+   while ((ent = apf_mesh->iterate(itr)))
+   {
+      // Get vertices
+      apf::Downward verts;
+      // int num_vert =
+      apf_mesh->getDownward(ent,0,verts);
+      // Get attribute Tag vs Geometry
+      int attr = 1;
+      /*if (apf_mesh->hasTag(ent,atts)){
```

Comment:
@KazemKamran: can you please clean up commented/unused code like this before we merge?